media/multipart/src/main/java/io/helidon/media/multipart/MultiPartDecoder.java

```diff
@@ -106,120 +142,229 @@ public void onNext(DataChunk chunk) {
+                        // exit the parser iterator loop
+                        // the parser events processing will resume upon inner Subscriber demand
+                        return;
+                    case BODY:
+                        Iterator<BufferEntry> bodyIterator = event.asBodyEvent().body().iterator();
+                        bodyPartPublisher.nextIterator(bodyIterator);
+                        if (!bodyPartPublisher.drain()) {
+                            // the body was not fully drained, exit the parser iterator loop
+                            // the parser events processing will resume upon inner Subscriber demand
+                            return;
+                        }
+                        break;
+                    case END_PART:
+                        bodyPartPublisher.complete(null);
+                        bodyPartHeaderBuilder = null;
+                        bodyPartBuilder = null;
+                        bodyPartPublisher = null;
+                        requested = partsRequested.updateAndGet(v -> v == Long.MAX_VALUE || v < 0 ? v : v - 1);
+                        break;
+                    default:
+                }
             }
 
-            @Override
-            public void onNext(DataChunk item) {
-                item.release();
+            // we allow requested <= 0 to reach here, because we want to allow delivery of termination signals
+            // without requests or cancellations, but ultimately need to make sure we do not request from
+            // upstream, unless actual demand is observed (requested > 0)
+            if (requested < 0) {
+                if (cancelled) {
+                    upstream.cancel();
+                    cleanup();
+                    return;
+                }
+                // now is the right time to convert a bad request into an error
+                // bodyPartPublisher is null, so this error gets delivered only to outer Subscriber
+                error = new IllegalArgumentException("Expecting only positive requests for parts");
             }
 
-            @Override
-            public void onError(Throwable throwable) {
+            // ordering the delivery of errors after the delivery of all signals that precede it
+            // in the order of events emitted by the parser
+            if (error != null) {
+                if (bodyPartPublisher != null) {
+                    bodyPartPublisher.complete(error);
+                    bodyPartPublisher = null;
+                }
+                upstream.cancel();
+                downstream.onError(error);
+                cleanup();
+                return;
             }
 
-            @Override
-            public void onComplete() {
+            // parserIterator is drained, drop the reference to it, but keep it safe for any later invocations
+            parserIterator = EMPTY_PARSER_ITERATOR;
+            if (upstream == SubscriptionHelper.CANCELED) {
```

Comment:
There is a subtle race condition here. I'll propose a fix later.

(In essence, Dekker idiom is needed - `onError` concurrent with `drainBoth` executed by, say, `request` - may set `CANCELED` before `error`, so `drainBoth` may see `error` is `null`, but after reaching this line also see `CANCELED`, and deliver `onComplete`)

Reply:
(With the change so that `onError` does not set `upstream = CANCELED;` the race condition disappears)